test/NuGet.Core.Tests/NuGet.Protocol.Tests/HttpSource/HttpSourceAuthenticationHandlerTests.cs

```diff
@@ -312,6 +312,170 @@ public async Task SendAsync_WithWrongCredentials_StopsRetryingAfter3Times()
+            expectedCredentials = newCredentials;
+
+            // Setup GetCredentialsAsync mock to return new credentials.
+            credentialServiceMock
+                .Setup(x => x.GetCredentialsAsync(packageSource.SourceUri, It.IsAny<IWebProxy>(), CredentialRequestType.Unauthorized, It.IsAny<string>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(() => newCredentials);
+
+            // Setup TryGetLastKnownGoodCredentialsFromCache mock to return cached credentials for 8th request and new credentials afterwards.
+            credentialServiceMock
+                .Setup(x => x.TryGetLastKnownGoodCredentialsFromCache(packageSource.SourceUri, It.IsAny<bool>(), out It.Ref<ICredentials>.IsAny))
+                .Returns((Uri sourceUri, bool isProxyRequest, out ICredentials outCredentials) =>
+                {
+                    outCredentials = retryCount == 8 ? initialCredentials : newCredentials;
+                    return true;
+                });
+
+            // Send 2 more requests to the feed which succeed only if new credentials are used.
+            await SendAsync(client, 2);
+
+            Assert.Equal(11, retryCount);
+            credentialServiceMock.Verify(x => x.TryGetLastKnownGoodCredentialsFromCache(packageSource.SourceUri, It.IsAny<bool>(), out It.Ref<ICredentials>.IsAny), Times.Exactly(5));
+            credentialServiceMock.Verify(x => x.GetCredentialsAsync(packageSource.SourceUri, It.IsAny<IWebProxy>(), CredentialRequestType.Unauthorized, It.IsAny<string>(), It.IsAny<CancellationToken>()), Times.Exactly(2));
+
+            static HttpResponseMessage HandleRequest(int retryCount, ref bool isUnAuthorized, NetworkCredential credentials, NetworkCredential expectedCredentials)
+            {
+                if (isUnAuthorized)
+                {
+                    isUnAuthorized = false;
+                    return new HttpResponseMessage(HttpStatusCode.Unauthorized);
+                }
+                else
+                {
+                    isUnAuthorized = true;
+
+                    // Credentials are changed after 3 successful requests to the feed. Hence return 401 for 8th request.
+                    if (retryCount == 8)
+                    {
+                        isUnAuthorized = false;
+                        return new HttpResponseMessage(HttpStatusCode.Unauthorized);
+                    }
+
+                    Assert.Equal(expectedCredentials.UserName, credentials.UserName);
+                    Assert.Equal(expectedCredentials.Password, credentials.Password);
+                    return new HttpResponseMessage(HttpStatusCode.OK);
+                }
+            }
+        }
+
+        private static async Task SendAsync(HttpClient client, int count)
+        {
+            for (int i = 0; i < count; i++)
+            {
+                await client.SendAsync(new HttpRequestMessage(HttpMethod.Get, "http://foo"), CancellationToken.None);
```

Comment:
major nitpick (as in I realize this comment is not very impactful): it seems kind of weird that the tests use package source URL `http://package.source.test/`, but this request uses `http://foo`. It's this `HttpClient.SendAsync` call that is most important to use a domain name ending in `.test` to ensure it can't accidentally make HTTP requests over the real internet (if the `HttpClient` was configured with `HttpClientHandler` in its middleware, which this method doesn't know is true or false, and if the domain name actually resolves to an IP address). Also, trying to put myself in the mindset of an inexperienced dev, maybe there's a question if the request URL being different to the package source URL is a potential problem?

Anyway, definitely not worth changing if there's no other changes made. But if you end up pushing other changes, maybe consider changing this too. Not a big deal at all.

Reply:
There was a CI failure. Hence, I had to push a change to fix it. I made the following couple of changes in the latest commit.

1. Added a string constant for "http://package.source.test/" URL and referred it in all the tests.
2. Replaced all the references to "http://foo" with package source url constant added in this PR.

All the test passed locally. Hence pushed the changes.